README.md
# Node Inspector

[![Build Status](https://travis-ci.org/node-inspector/node-inspector.png?branch=master)](https://travis-ci.org/node-inspector/node-inspector)
[![Dependency Status](https://gemnasium.com/node-inspector/node-inspector.png)](https://gemnasium.com/node-inspector/node-inspector)
[![NPM version](https://badge.fury.io/js/node-inspector.png)](http://badge.fury.io/js/node-inspector)


## Overview

Node Inspector is a debugger interface for Node.js applications that uses the
Blink Developer Tools (formerly WebKit Web Inspector).

The project maintenance and support is sponsored by
[StrongLoop](http://strongloop.com).

### Table of Content

 * [Quick Start](#quick-start)
 * [Features](#features)
 * [Known Issues](#known-issues)
 * [Troubleshooting](#troubleshooting)
 * [Advanced Use](#advanced-use)
 * [Configuration](#configuration)
 * [Contributing](#contributing-code)
 * [Credits](#credits)

## Quick Start

#### Install

```sh
$ npm install -g node-inspector
```

#### Start

```sh
$ node-debug app.js
```

where ```app.js``` is the name of your main Node application JavaScript file.

#### Debug

The `node-debug` command will load Node Inspector in your default browser.

> **NOTE:** Node Inspector works in Chrome and Opera only. You have to re-open
the inspector page in one of those browsers if another browser
is your default web browser (e.g. Safari or Internet Explorer).

Node Inspector works almost exactly as the Chrome Developer Tools. Read the
excellent
[DevTools overview](http://code.google.com/chrome/devtools/docs/scripts.html)
to get started.

Other useful resources:
 - Documentation specific to Node Inspector provided by StrongLoop:
  [Running Node Inspector](http://docs.strongloop.com/display/DOC/Running+Node+Inspector)
 - Miroslav's talk
  [How to Debug Node Apps with Node Inspector](https://vimeo.com/77870960)
 - Danny's [screencasts](http://www.youtube.com/view_play_list?p=A5216AC29A41EFA8)
   (most likely outdated by now)
 - [Getting Started from scratch](http://github.com/node-inspector/node-inspector/wiki/Getting-Started---from-scratch)
   on wiki (most likely outdated by now)

## Features

The Blink DevTools debugger is a powerful JavaScript debugger interface.
Node Inspector supports almost all of the debugging features of DevTools, including:

* Navigate in your source files
* Set breakpoints (and specify trigger conditions)
* Step over, step in, step out, resume (continue)
* Inspect scopes, variables, object properties
* Hover your mouse over an expression in your source to display its value in
  a tooltip
* Edit variables and object properties
* Continue to location
* Break on exceptions
* Disable/enable all breakpoints

### Cool stuff
* Node Inspector uses WebSockets, so no polling for breaks.
* Remote debugging
* [Live edit of running code](http://github.com/node-inspector/node-inspector/wiki/LiveEdit),
  optionally persisting changes back to the file-system.
* Set breakpoints in files that are not loaded into V8 yet - useful for
  debugging module loading/initialization.
* Javascript from top to bottom :)
* Embeddable in other applications - see [Embedding HOWTO](docs/embedding.md)
  for more details.

## Known Issues

* Be careful about viewing the contents of Buffer objects,
  each byte is displayed as an individual array element;
  for most Buffers this will take too long to render.
* While not stopped at a breakpoint the console doesn't always
  behave as you might expect. See the
  [issue #146](https://github.com/node-inspector/node-inspector/issues/146).
* Profiler is not implemented yet. Have a look at
  [node-webkit-agent](https://github.com/c4milo/node-webkit-agent)
  in the meantime.
* Break on uncaught exceptions does not work in all Node versions,
  you need at least v0.11.3 (see
  [node#5713](https://github.com/joyent/node/pull/5713)).
* Debugging multiple processes (e.g. cluster) is cumbersome.
  Read the following blog post for instructions:
  [Debugging Clustered Apps with Node-Inspector](http://strongloop.com/strongblog/whats-new-nodejs-v0-12-debugging-clusters/)

## Troubleshooting

#### My script runs too fast to attach the debugger.

The debugged process must be started with `--debug-brk`, this way the script is paused on the first line.

Note: `node-debug` adds this option for you by default.

#### I got the UI in a weird state.

When in doubt, refresh the page in browser

#### Can I debug remotely?

Yes. Node Inspector must be running on the same machine, but your browser can be anywhere.
Just make sure port 8080 is accessible.

#### How do I specify files to hide?

Create a JSON-encoded array.  You must escape quote characters when using a command-line option.

```sh
$ node-inspector --hidden='["node_modules/framework"]'
```

Note that the array items are interpreted as regular expressions.

#### UI doesn't load or doesn't work and refresh didn't help

Make sure that you have adblock disabled as well as any other content blocking scripts and plugins.

#### How can I (selectively) delete debug session metadata?

You may want to delete debug session metadata if for example Node Inspector gets in a bad state with some
watch variables that were function calls (possibly into some special c-bindings).  In such cases, even restarting
the application/debug session may not fix the problem.

Node Inspector stores debug session metadata in the HTML5 local storage.
You can inspect the contents of local storage and remove any items as
needed. In Google Chrome, you can execute any of the following in the JavaScript console:

```js
// Remove all
window.localStorage.clear()
// Or, to list keys so you can selectively remove them with removeItem()
window.localStorage
// Remove all the watch expressions
window.localStorage.removeItem('watchExpressions')
// Remove all the breakpoints
window.localStorage.removeItem('breakpoints')
```

When you are done cleaning up, hit refresh in the browser.

#### Node Inspector takes a long time to start up.

Try setting --no-preload to true. This option disables searching disk for *.js at startup.

#### How do I debug Mocha unit-tests?

You have to start `_mocha` as the debugged process and make sure
the execution pauses on the first line. This way you have enough
time to set your breakpoints before the tests are run.

```sh
$ node-debug _mocha
```

## Advanced Use

While running `node-debug` is a convenient way to start your debugging
session, there may come time when you need to tweak the default setup.

There are three steps needed to get you up and debugging:

#### 1. Start the Node Inspector server

```sh
$ node-inspector
```

You can leave the server running in background, it's possible to debug
multiple processes using the same server instance.

#### 2. Enable debug mode in your Node process

You can either start Node with a debug flag like:

```sh
$ node --debug your/node/program.js
```

or, to pause your script on the first line:

```sh
$ node --debug-brk your/short/node/script.js
```

Or you can enable debugging on a node that is already running by sending
it a signal:

1. Get the PID of the node process using your favorite method.
`pgrep` or `ps -ef` are good

    ```sh
    $ pgrep -l node
    2345 node your/node/server.js
    ```

2. Send it the USR1 signal

    ```sh
    $ kill -s USR1 2345
    ```

##### Windows

Windows does not support UNIX signals. To enable debugging, you can use
an undocumented API function `process._debugProcess(pid)`:

1. Get the PID of the node process using your favorite method, e.g.

    ```sh
    > tasklist /FI "IMAGENAME eq node.exe"

    Image Name                     PID Session Name        Session#    Mem Usage
    ========================= ======== ================ =========== ============
    node.exe                      3084 Console                    1     11,964 K
    ```

2. Call the API:

    ```sh
    > node -e "process._debugProcess(3084)"
    ```

#### 3. Load the debugger UI

Open http://127.0.0.1:8080/debug?port=5858 in the Chrome browser.

## Configuration

### node-debug

Command line options:

```
--debug-brk, -b         Break on the first line (`node --debug-brk`) [default: true]
--web-port, -p, --port  Node Inspector port (`node-inspector --web-port={port}`)
--debug-port, -d        Node/V8 debugger port (`node --debug={port}`)
--cli, -c               CLI mode, do not open browser.
--version, -v           Print Node Inspector's version.
--help, -h              Show this help.
```

### node-inspector

node-inspector uses [rc](https://npmjs.org/package/rc)
[[github]](https://github.com/dominictarr/rc) module to collect options.

Places for configuration:
* command line arguments (parsed by optimist)
* enviroment variables prefixed with ```node-inspector_```
* if you passed an option ```--config file``` then from that file
* a local ```.node-inspectorrc``` or the first found looking in ```./ ../ ../../
 ../../../``` etc.
* ```$HOME/.node-inspectorrc```
* ```$HOME/.node-inspector/config```
* ```$HOME/.config/node-inspector```
* ```$HOME/.config/node-inspector/config```
* ```/etc/node-inspectorrc```
* ```/etc/node-inspector/config```
* options from ```config.json``` for backward compatibility
* defaults described in Node Inspector`s [./lib/config.js](lib/config.js).

All configuration sources that where found will be flattened into one object,
so that sources earlier in this list override later ones.

Use dashed option names in RC files. Sample config file:
```js
{
  "web-port": 8088,
  "web-host": null,
  "debug-port": 5858,
  "debug-host": null,
  "save-live-edit": true,
  "no-preload": true,
  "hidden": [],
  "stack-trace-limit": 50
}
```

List of predefined options:
```
       Option            Default                  Description
--help               |             | Print information about options
--web-port           |    8080     | Port to host the inspector
--web-host           |  127.0.0.1  | Host to listen on
--debug-port         |    5858     | Port to connect to the debugging app
--debug-host         |  127.0.0.1  | Host where the debugged app is running
--save-live-edit     |    false    | Save live edit changes to disk
                     |             |   (update the edited files)
--preload            |    true     | Preload *.js files. You can disable this option
                     |             |    to speed up the startup.
--hidden             |     []      | Array of files to hide from the UI
                     |             |   (breakpoints in these files will be ignored)
--stack-trace-limit  |     50      | Number of stack frames to show on a breakpoint
```

## Contributing Code

Making Node Inspector the best debugger for node.js cannot be achieved without
the help of the community. The following resources should help you to get
started.

* [Contributing](https://github.com/node-inspector/node-inspector/wiki/Contributing)
* [Developer's Guide](https://github.com/node-inspector/node-inspector/wiki/Developer%27s-Guide)
* [Easy Picks](https://github.com/node-inspector/node-inspector/issues?direction=asc&labels=Easy+Pick&page=1&sort=updated&state=open)

## Credits

Maintainers
 - [Danny Coates](https://github.com/dannycoates) - the original author
   and a sole maintainer for several years.
 - [Miroslav Bajtoš](https://github.com/bajtos) - the current maintainer,
   sponsored by [StrongLoop](http://strongloop.com).

Big thanks to the many contributors to the project, see [AUTHORS](AUTHORS).


ChangeLog
2016-12-21, Version 0.7.5

* Add support for remote debugging (#956) (Junil Kim)


2014-06-05, Version 0.7.4

* Non-ascii string saving fix (junecoder)

* front-end: fix for the Safari browser (ivan baktsheev)

* node-debug: Detecting Windows CMD files (3y3)

* node-debug: Passing NodeJS options to debugged process (3y3)

* node-debug: Fixed passing inspector's 'no-' arg in node-debug (3y3)

* Fix conversion of config options (3y3)


2014-04-08, Version 0.7.3

* ScriptFileStorage: fix fs.readFile usage on v0.8 (Miroslav Bajtoš)

* Update dependencies (Miroslav Bajtoš)

* Switch to core _debugger module. (3y3)


2014-03-21, Version 0.7.2

* Fixed a race condition in the debug-server.js (Vasil Dininski)

* Use jshint instead of gjslint (Miroslav Bajtoš)

* fixjsstyle cleanup (3y3)

* debugger isRunning state changes (3y3)


2014-03-11, Version 0.7.1

* Added subtype property to RemoteObject (3y3)

* Checked Invalid Date object (3y3)

* Fixed wrong proto object link (3y3)

* Checking configurable option (3y3)

* Properties conversion isolated to function (3y3)

* Handle shebang in ScriptFileStorage load/save (Miroslav Bajtoš)

* Implement Debugger.setSkipAllPauses (Miroslav Bajtoš)

* Fix links (Miroslav Bajtoš)

* MAINTAINERS: use imperative sentence style (Miroslav Bajtoš)


2014-02-20, Version 0.7.0

* README restructuralization (Miroslav Bajtoš)

* Rename readme.md to README.md (Miroslav Bajtoš)

* Update AUTHORS (Miroslav Bajtoš)

* Defer "scriptParsed" until "Page.getResourceTree" (Miroslav Bajtoš)

* Lower the frequency of auto reloads (Miroslav Bajtoš)

* Update .npmignore (Miroslav Bajtoš)

* Clean up node-debug help message (Sam Roberts)

* bin/inspector: Add -h -v --version options (Miroslav Bajtoš)

* Speed up initial parsing of source-map URLs (Miroslav Bajtoš)

* Cache results of ScriptFileStorage.listScripts. (Miroslav Bajtoš)

* Ignore EACCESS errors when listing sources (Miroslav Bajtoš)

* Tighten findAllApplicationScripts heuristics (Miroslav Bajtoš)

* Implement node-debug. (Miroslav Bajtoš)

* RuntimeAgent: implement releaseObject() as no-op (Miroslav Bajtoš)

* Fix DebuggerAgent.setVariableValue to object (Miroslav Bajtoš)

* Add WebSocketServer error handler. (Miroslav Bajtoš)

* README: added shell code higlight (Ionică Bizău)

* Fix conversion of Error objects (Miroslav Bajtoš)

* Use WebSockets instead of socket.io. (Kenneth Auchenberg)

* Small indentation fixes (3y3)

* Help message format update (3y3)

* Reload browser when detached from target (ChrisWren)

* Make socket.io work with subfolder hosting. (Pritam Baral)


2014-01-23, Version 0.7.0-2

* Updated AUTHORS. (Miroslav Bajtoš)

* Fixed display of RegExp objects. (3y3)

* Fixed DebuggerClient.resume acting like stepInto (Miroslav Bajtoš)

* Fixed timing-dependent unit-tests failures (Miroslav Bajtoš)

* test: improved describe/it names (Miroslav Bajtoš)

* Fixed misspelled word "Visit" in docs (Peter Lyons)

* Fix Date display format. Issue #281 (3y3)

* Fixed throw on DebuggerAgent.resume() (Lennon Pulda-Grealy)

* Added large String support for debugger evaluation requests (Peter Flannery)


2013-12-26, Version 0.7.0-1

* Improved formatting of Date objects in inspector (Sergey Krilov)

* Added --no-preload to disable glob of all *.js (Dick Hardt)

* Update readme with faq about browser storage and debug metadata (Karan Batra-Daitch)

* Fixed CallFramesProvider test on Node v0.11. (Miroslav Bajtos)


2013-12-04, Version 0.6.2

* Updated AUTHORS. (Miroslav Bajtos)

* docs: explain dashed name rc config conversion (cattail)

* Removed false error messages on reload (Dave)


2013-11-17, Version 0.6.1

* Added debug logging of communication (Miroslav Bajtos)


2013-11-12, Version 0.6.0

* gjslint: removed workaround for linter bug (Miroslav Bajtos)

* Updated AUTHORS. (Miroslav Bajtos)

* Added `stackTraceLimit` configuration, fixes #96. (ssafejava)

* test: fixed launcher errors on windows (Miroslav Bajtos)

* Remove dependence on HTTP for socket io connection to server (Matthew O'Riordan)

* README: added WTF entry about adblock (Gary Katsevman)

* README: howto signal SIGUSR1 on windows (Miroslav Bajtos)

* lib,test: fixed lint issues (Miroslav Bajtos)


2013-09-20, Version 0.5.0

* Quick fix for properties displayed twice in UI (Miroslav Bajtos)

* README: added FAQ entry for --hidden option (Miroslav Bajtos)

* lib: fixed hiding of script files (Miroslav Bajtos)

* lib: fix config handling (Miroslav Bajtos)

* Updated AUTHORS. (Miroslav Bajtos)

* Fixed wrong hint in error message if web port is used. (Adam Hořčica)

* NetworkAgent: support data scheme URLs (Miroslav Bajtos)

* package.json: upgraded to express 3.4 (Miroslav Bajtos)

* lib,Overrides: support the updated front-end (Miroslav Bajtos)

* front-end: patched inspector.html (Miroslav Bajtos)

* front-end: upgraded to Blink branch 1625, r157454 (Miroslav Bajtos)

* tools: implemented update-front-end.sh (Miroslav Bajtos)


2013-09-05, Version 0.4.0

* Emit 'listening' and 'error' events (Miroslav Bajtos)

* README: fixed headings level (Miroslav Bajtos)

* New configuration system based on RC module (3y3)


2013-08-21, Version 0.3.4

* Bugfix: crash when breakpoint was hit in eval()-ed (Miroslav Bajtos)


2013-08-12, Version 0.3.3

* package.json: upgraded to express 3.3 (Miroslav Bajtos)

* README: added section 'Contributing Code' (Miroslav Bajtos)

* README: Updated "Thanks" section (Miroslav Bajtos)

* Updated AUTHORS list. (Miroslav Bajtos)

* DebugServer: Fix for source map paths on win (pflannery)

* README: added gemnasium dependency status badge (Miroslav Bajtos)

* package.json: added `lint` script (Miroslav Bajtos)

* README: link to issue #146 (Miroslav Bajtoš)

* PageAgent: improved main-file detection (Miroslav Bajtos)


2013-07-22, Version 0.3.2

* Overrides: show main app file (Miroslav Bajtos)

* PageAgent: fix for main script without extension (Miroslav Bajtos)


2013-07-15, Version 0.3.1


lib/config.js
var rc = require('rc'),
    util = require('util');

var conversions = {
  checkIfNull: function(value) {
    return value && value !== 'null' ? value : null;
  },
  keyToCamelKey: function(value) {
    return value.replace(/-(.)/g, function(_, lower) {
      return lower.toUpperCase();
    });
  },
  rcToInnerConfig: function(rcConfig) {
    var options = {};
    Object.keys(rcConfig).forEach(function(key) {
      var camelKey = conversions.keyToCamelKey(key),
          fixedVal = rcConfig[key],
          predefined;

      predefined = !!definitions[key];

      if (predefined) {
        try {
          fixedVal = definitions[key].convert(fixedVal);
        }
        catch (e) {
          console.warn('Cannot convert config option %s: %s.', key, e.message || e);
        }
      }
      options[camelKey] = fixedVal;
    });
    return options;
  },
  stringToArray: function(value) {
    var hidden;
    if (typeof value === 'string') {
      try {
        value = JSON.parse(value);
      } catch (e) {
        throw new Error('The value is not a valid JSON. ' + (e.message || e));
      }
    }
    if (util.isArray(value)) {
      hidden = value.map(function(s) { return new RegExp(s, 'i'); });
    } else {
      var msg = 'The value ' + JSON.stringify(value) + ' is not an array.';
      throw new Error(msg);
    }
    return hidden;
  },
  stringToBoolean: function(value) {
    return !!value;
  },
  stringToInt: function(value) {
    return parseInt(value, 10);
  }
};
var definitions = {
  'help': {
    desc: 'Show this help',
    convert: conversions.stringToBoolean,
    defaultValue: false
  },
  'version': {
    desc: 'Print Node Inspector\'s version',
    convert: conversions.stringToBoolean,
    defaultValue: false
  },
  'web-port': {
    desc: 'Port to host the inspector',
    convert: conversions.stringToInt,
    defaultValue: 8080
  },
  'web-host': {
    desc: 'Host to listen on',
    convert: conversions.checkIfNull,
    defaultValue: ''
  },
  'debug-port': {
    desc: 'Port to connect to the debugging app',
    convert: conversions.stringToInt,
    defaultValue: 5858
  },
  'debug-host': {
    desc: 'Host where the debugged app is running',
    convert: conversions.checkIfNull,
    defaultValue: ''
  },
  'save-live-edit': {
    desc: 'Save live edit changes to disk (update the edited files)',
    convert: conversions.stringToBoolean,
    defaultValue: false
  },
  'preload': {
    desc: 'Preload *.js files. You can disable this option to speed up the startup.\n' +
          '    (command-line parameter: \u001b[92m--no-preload\u001b[0m)',
    convert: conversions.stringToBoolean,
    defaultValue: true
  },
  'hidden': {
    desc: 'Array of files to hide from the UI (breakpoints in these files' +
          ' will be ignored)',
    convert: conversions.stringToArray,
    defaultValue: []
  },
  'stack-trace-limit': {
    desc: 'Number of stack frames to show on a breakpoint',
    convert: conversions.stringToInt,
    defaultValue: 50
  }
};

var defaults = collectDefaultsFromDefinitions();
var rcConfig = rc('node-inspector', defaults);
var config = conversions.rcToInnerConfig(rcConfig);

if (config.noPreload !== undefined) {
  // Deprecated in v0.7.3
  console.warn('The config option `no-preload` is deprecated, use `preload` instead');
  config.preload = config.preload || !config.noPreload;
}

module.exports = config;

module.exports._collectDefaults = function() {
  var dashedKeyDefaults = collectDefaultsFromDefinitions();
  return conversions.rcToInnerConfig(dashedKeyDefaults);
};

module.exports._describeOptions = function() {
  return Object.keys(definitions)
    .map(function constructMessagePart(key) {
      var definition = definitions[key];

      var defaultValue = definition.defaultValue;
      var defaultString = JSON.stringify(definition.defaultValue);

      var typeString = Object.prototype.toString.call(defaultValue);
      var matchedType = /^\[object (.*)\]$/.exec(typeString)[1];

      var optionKey = '\u001b[92m--' + key;
      var optionTypeAndDefault =
        matchedType !== 'Undefined' && matchedType !== 'Boolean' ?
          '=\u001b[90m{' + matchedType + '}' +
            ' \u001b[96m(default: ' + defaultString + ')' :
          '';
      var optionDescription = '\u001b[0m' + definition.desc;

      return '    ' + optionKey + optionTypeAndDefault +
        '\n    ' + optionDescription;
    })
    .join('\n\n');
};

function collectDefaultsFromDefinitions() {
  var options = {};

  Object.keys(definitions).forEach(function(key) {
    options[key] = definitions[key].defaultValue;
  });

  return options;
}
